src/resources/resource_storage.rs

```diff
@@ -10,14 +10,124 @@ use thiserror::Error;
 use super::{PermissionMask, Resource, ResourceType};
 
 /// Unified resource storage for both redirects and scriptlets.
-#[derive(Default)]
+///
+/// By default, this uses an in-memory storage implementation, however this can be changed using
+/// a custom [ResourceStorageBackend] if desired.
 pub struct ResourceStorage {
+    #[cfg(not(feature = "single-thread"))]
+    backend: Box<dyn ResourceStorageBackend + Sync + Send>,
```

Comment:
hm, I'm not strongly again `Box<dyn ..>` it makes the code less clear and may have some negative perf effect (if it's a hot path)